zebra/zebra_rnh.c

```diff
@@ -1268,73 +1270,181 @@ int zebra_send_rnh_update(struct rnh *rnh, struct zserv *client,
-		vty_out(vty, " %s",
-			inet_ntop(AF_INET6, &nexthop->gate.ipv6, buf, BUFSIZ));
-		if (nexthop->ifindex)
-			vty_out(vty, ", via %s",
-				ifindex2ifname_per_ns(zns, nexthop->ifindex));
+		if (json) {
+			json_object_string_addf(json, "ip", "%pI6",
+						&nexthop->gate.ipv6);
+			if (nexthop->ifindex)
+				json_object_string_add(
+					json, "interface",
+					ifindex2ifname_per_ns(
+						zns, nexthop->ifindex));
+		} else {
+			vty_out(vty, " %s",
+				inet_ntop(AF_INET6, &nexthop->gate.ipv6, buf,
+					  BUFSIZ));
+			if (nexthop->ifindex)
+				vty_out(vty, ", via %s",
+					ifindex2ifname_per_ns(
+						zns, nexthop->ifindex));
+		}
 		break;
 	case NEXTHOP_TYPE_IFINDEX:
-		vty_out(vty, " is directly connected, %s",
-			ifindex2ifname_per_ns(zns, nexthop->ifindex));
+		if (json) {
+			json_object_string_add(
+				json, "interface",
+				ifindex2ifname_per_ns(zns, nexthop->ifindex));
+			json_object_boolean_true_add(json, "directlyConnected");
+		} else {
+			vty_out(vty, " is directly connected, %s",
+				ifindex2ifname_per_ns(zns, nexthop->ifindex));
+		}
 		break;
 	case NEXTHOP_TYPE_BLACKHOLE:
-		vty_out(vty, " is directly connected, Null0");
+		if (json) {
+			json_object_string_add(json, "interface", "Null0");
+			json_object_boolean_true_add(json, "directlyConnected");
+		} else {
+			vty_out(vty, " is directly connected, Null0");
+		}
 		break;
 	default:
 		break;
 	}
-	vty_out(vty, "\n");
+
+	if (!json)
+		vty_out(vty, "\n");
 }
 
-static void print_rnh(struct route_node *rn, struct vty *vty)
+static void print_rnh(struct route_node *rn, struct vty *vty, json_object *json)
 {
 	struct rnh *rnh;
 	struct nexthop *nexthop;
 	struct listnode *node;
 	struct zserv *client;
 	char buf[BUFSIZ];
+	json_object *json_nht = NULL;
+	json_object *json_client_array = NULL;
+	json_object *json_client = NULL;
+	json_object *json_nexthop_array = NULL;
+	json_object *json_nexthop = NULL;
 
 	rnh = rn->info;
-	vty_out(vty, "%s%s\n",
-		inet_ntop(rn->p.family, &rn->p.u.prefix, buf, BUFSIZ),
-		CHECK_FLAG(rnh->flags, ZEBRA_NHT_CONNECTED) ? "(Connected)"
-							    : "");
-	if (rnh->state) {
-		vty_out(vty, " resolved via %s\n",
-			zebra_route_string(rnh->state->type));
-		for (nexthop = rnh->state->nhe->nhg.nexthop; nexthop;
-		     nexthop = nexthop->next)
-			print_nh(nexthop, vty);
-	} else
-		vty_out(vty, " unresolved%s\n",
+
+	if (json) {
+		json_nht = json_object_new_object();
+		json_nexthop_array = json_object_new_array();
+		json_client_array = json_object_new_array();
+
+		json_object_object_add(
+			json,
+			inet_ntop(rn->p.family, &rn->p.u.prefix, buf, BUFSIZ),
+			json_nht);
+		json_object_boolean_add(
+			json_nht, "nhtConnected",
+			CHECK_FLAG(rnh->flags, ZEBRA_NHT_CONNECTED));
+		json_object_object_add(json_nht, "clientList",
+				       json_client_array);
+		json_object_object_add(json_nht, "gates", json_nexthop_array);
+	} else {
+		vty_out(vty, "%s%s\n",
+			inet_ntop(rn->p.family, &rn->p.u.prefix, buf, BUFSIZ),
```

Comment:
I think using the formatting is preferred instead of `inet_ntop`. It would be nice to get rid of the buffer sized with `BUFSIZ` too (if I recall correctly it is 4096 which is big).

Reply:
@rzalamena which format to use to represent prefix in address form (both ipv4/ipv6).  The `rn->p` is in prefix form but nht expect to display only address part and not prefixlen. 

```
tor-1# show ipv6 nht vrf default
fe80::202:ff:fe00:2b(Connected) <<<<<<< if %pFX is used then `fe80::202:ff:fe00:2b/128` is displayed
 resolved via connected
```

Reply:
There is not printfrr extension to distinguish by AF right now. I think it's fine here.